compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2457,25 +2473,31 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       val casesEmptyBody2 = typedCases(casesEmptyBody1, EmptyTree, defn.ThrowableType, WildcardType)
       val expr1 = typed(addCanThrowCapabilities(tree.expr, casesEmptyBody2), pt.dropIfProto)
 
-      // Since we don't know at which point the the exception is thrown in the body,
-      // we have to collect any reference that is once retracted.
-      nnInfo = expr1.notNullInfo.retractedInfo
-
-      val casesCtx = ctx.addNotNullInfo(nnInfo)
+      val casesCtx = ctx.addNotNullInfo(expr1.notNullInfo.retractedInfo)
       val cases1 = typedCases(tree.cases, EmptyTree, defn.ThrowableType, pt.dropIfProto)(using casesCtx)
       expr1 :: cases1
     }: @unchecked
     val cases2 = cases2x.asInstanceOf[List[CaseDef]]
-
+    val tryNNInfo = expr2.notNullInfo
     // It is possible to have non-exhaustive cases, and some exceptions are thrown and not caught.
     // Therefore, the code in the finalizer and after the try block can only rely on the retracted
     // info from the cases' body.
-    if cases2.nonEmpty then
-      nnInfo = nnInfo.seq(cases2.map(_.notNullInfo.retractedInfo).reduce(_.alt(_)))
-
-    val finalizer1 = typed(tree.finalizer, defn.UnitType)(using ctx.addNotNullInfo(nnInfo))
-    nnInfo = nnInfo.seq(finalizer1.notNullInfo)
-    assignType(cpy.Try(tree)(expr2, cases2, finalizer1), expr2, cases2).withNotNullInfo(nnInfo)
+    val catchNNInfo = if cases2.nonEmpty then
+      tryNNInfo.seq(cases2.map(_.notNullInfo).reduce(_.alt(_)))
+    else
+      NotNullInfo.empty
+    val normalResolveNNInfo = tryNNInfo.alt(catchNNInfo)
+    val exceptionalResolveNNInfo = catchNNInfo.retractedInfo
+
+    val finalizer1 = typed(tree.finalizer, defn.UnitType)(using ctx.addNotNullInfo(exceptionalResolveNNInfo))
+    val normalFinalNNInfo = normalResolveNNInfo.seq(finalizer1.notNullInfo)
+    val exceptionalFinalNNInfo = exceptionalResolveNNInfo.seq(finalizer1.notNullInfo)
+    val resNNInfo =
+      if (cases2.forall(_.tpe == defn.NothingType)) then
+        normalFinalNNInfo
```

Comment:
Also, I would like to name the situation where all cases terminates exceptionally as "exceptional"..